### ISSUE TITLE
error types

### ISSUE THREAD

**Issue by andre-merzky** [radical-cybertools/radical.entk#299]
Why does EnTK define its own error types when the types already exist in Python (like `ValueError`)?  That makes it cumbersome to use, and needs extra documentation where none should be needed.  In fact,  I don't think any EnTK error type which needs its own definition.  No custom error is used in any example nor in the documentation...

```
re.TypeError -> TypeError
re.ValueError -> ValueError
MissingError -> AttributeError
ExistsError ->  not used in EnTK
MatchError ->  not used in EnTK
```

**Comment by andre-merzky** [comment]
What is the resolution on this?

**Comment by vivek-bala** [comment]
Yea, this was not to be closed. Needs more discussion. Apologies. 

I removed the 2 error types not used in EnTK. I have kept the other three
to keep the error message meaningful, for e.g., not only do we want to
say that the type is incorrect, we also want to say what the expected
type is.

I believe Python complains if I try to overload the default exception types.

What do you think? Let me know if I am missing something.

**Comment by andre-merzky** [comment]
You don't need to overload them - you can raise them with useful messages:

```
raise ValueError('negative temperatures in K are undefined')
```
What is a derived or custom type adding?

The reason why I am asking is that we did the same in SAGA - but by now I find those custom types cumbersome to use, and and am reconsidering if they are worth the trouble...

**Comment by vivek-bala** [comment]
If we had a large number of ValueErrors and we wanted to change the error message to make it more
meaningful (e.g. ''negative temperatures in K are undefined, K can only be between (0,1000)'' and different error messages for different attributes), we would have to change all ValueErrors individually.

I guess my goal here was to use a function where I can specify the incorrect attribute, the incorrect value, expected value(s) (see [here](https://github.com/radical-cybertools/radical.entk/blob/master/src/radical/entk/exceptions.py#L39)). The path I chose was custom exceptions.

Another way I can think of is to create a function: ```raise_value_error(obj, attribute, expected_value, actual_value)``` which raises ValueError with the appropriate error message. In EnTK, instead of raising an exception, we call the raise_value_error function. What do you think?

**Comment by andre-merzky** [comment]
I don't see why that wouldn't work with inherited exceptions.  But I would also like to point out that type checking is something Python does for you already, much quicker, and the exception message can be almost as useful (see option 3):

```
#!/usr/bin/env python

i =  1     # the type we want
s = 'one'  # what the user specifies


# ------------------------------------------------------------------------------
#
# Option 1
#
class MyValueError(ValueError):

    def __init__(self, msg, expected, found):

        msg = '%s: expected %s, found %s' % (msg, expected, found)

        ValueError.__init__(self, msg)


try:
    if type(i) != type(s):
        raise MyValueError('check test 1', expected=type(s), found=type(i))
except ValueError as e:
    print 'caught exception: %s' % str(e)
  # raise


# ------------------------------------------------------------------------------
#
#  Option 2
#
class MyValueCheck(ValueError):

    def __init__ (self, msg, data, expected):

        if not isinstance(expected, list):
            expected = [expected]

        if type(data) not in [type(e) for e in expected]:
            msg = '%s: expected %s, found %s' \
                % (msg, [type(e) for e in expected], type(data))

            ValueError.__init__(self, msg)
            raise self


try:
    MyValueCheck('check test 2', i, s)

except ValueError as e:
    print 'caught exception: %s' % str(e)
  # raise


# ------------------------------------------------------------------------------
#
#  Option 3 - the Python way
#
try:
    # make sure we have an int.  We also accept the string "1"!
    s = int(s)  # NOTE: we expect s to be an integer (number of sausages)

except ValueError as e:
    print 'caught exception: %s' % str(e)
    raise

# ------------------------------------------------------------------------------
```

**Comment by vivek-bala** [comment]
Thanks for the examples Andre.
Please see the output of the three options when 'i' is an integer:
```
caught exception: check test 1: expected <type 'str'>, found <type 'int'>
caught exception: check test 2: expected [<type 'str'>], found <type 'int'>
caught exception: invalid literal for int() with base 10: 'one'
Traceback (most recent call last):
  File "test.py", line 52, in <module>
    s = int(s)  # NOTE: we expect s to be an integer (number of sausages)
ValueError: invalid literal for int() with base 10: 'one'
```

Output of the three when 'i' is a dict:
```
caught exception: check test 1: expected <type 'str'>, found <type 'dict'>
caught exception: check test 2: expected [<type 'str'>], found <type 'dict'>
caught exception: dictionary update sequence element #0 has length 1; 2 is required
Traceback (most recent call last):
  File "test.py", line 52, in <module>
    s = dict(s)  # NOTE: we expect s to be an integer (number of sausages)
ValueError: dictionary update sequence element #0 has length 1; 2 is required
```

I think options 1 and 2 are quite explanatory on what the issue is. I am more 
confident that if the user looks at the output of options 1 or 2, the user will 
be able to fix the issue before requiring our help.

FWIW I am not disagreeing with you that we are creating and maintaining 
another set of exceptions. I think that is worth it though. 

What do you think?

**Comment by andre-merzky** [comment]
I am not really opposed to specific EnTK error types!  The ticket is about me not being able to except for `TypeError`, because the EnTK error types all inherit from the base Exception class, not from their respective Python equivalents.  EnTK (basically) inherits like this:

```python
re.TypeError -> Exception
re.ValueError -> Exception 
re.MissingError -> Exception
re.ExistsError ->  Exception
re.MatchError ->  Exception
```

and I propose to inherit like this:

```python
re.TypeError -> TypeError
re.ValueError -> ValueError
re.MissingError -> AttributeError
re.ExistsError ->  not used in EnTK
re.MatchError ->  not used in EnTK
```

I also think that EnTK should be more accepting of types, and that the error checking code can be simplified (see example 2) - but that's not part of this ticket :-)

**Comment by vivek-bala** [comment]
Ah okay, sorry I missed that! I agree with you on that!

**Comment by andre-merzky** [comment]
Yeah, now that I reread my first message, this was not really clear, apologies.  Maybe it wasn't clear to me either ;-)

**Comment by mturilli** [comment]
Even if it is not specifically related to the original (or not :)) goal of the ticket, I would like to keep going the discussion about error messages. As we know, due to the distributed nature of our tools and that of our users (students, domain scientists not necessarily expert coders) error messages are particularly important. I think Vivek's analysis is the right way to evaluate and improve upon our error messages. Should we open a ticket for that?

**Comment by vivek-bala** [comment]
Changed the inheritance to:
```python
re.TypeError -> TypeError
re.ValueError -> ValueError
re.MissingError -> AttributeError
```

Thanks Andre. Released in 0.7.14.
